content/technotes/2020-07-07-github-pat-credentials.md

```diff
@@ -0,0 +1,99 @@
+tags:
+  - R
+  - package
+  - gert
+  - credentials
+  - git
+  - github
+---
+
+We have been working hard behind the scenes on the upcoming release of our new git package named [gert](https://docs.ropensci.org/gert), a joint effort from rOpenSci and the Tidyverse team. One of the main features of gert is the out-of-the-box authentication mechanism, which is provided via the new [credentials](https://docs.ropensci.org/credentials/articles/intro.html) package.
+
+Among other things, the credentials package makes it possible to save and load https authentication details from the [git credential store](https://git-scm.com/docs/git-credential-store), which is part of the official command-line git. Thereby credentials are automatically shared between command line git and the gert package, while safely stored by your operating system's preferred password manager.
+
+In this post we show how you can take this one step further, and use the credentials package to save your `GITHUB_PAT` in the git credential store. This way you can authenticate with the GitHub API using the same token that is used for HTTPS remotes in git and gert. This is convenient for users, and also provides package authors with a mechanism to prompt the user for credentials, without having to take responsibility for managing tokens.
+
+
+## What is a Personal Access Token
+
+GitHub allows you to generate [Personal Access Tokens](https://github.com/settings/tokens), which you can use instead of your password when authenticating over HTTPS, both for git remotes and the GitHub API. Major advantages are:
+
+ - If you have enabled 2FA, you must use a PAT to authenticate programmatically. You have no choice.
+ - A PAT can easily be revoked or updated. You can even automate this via the Github API.
+ - You can generate many PATs with specific permissions, giving you fine-grained security control.
+
+In conclusion, if you are a responsible GitHub user, you have enabled 2FA on your account, and you only ever enter your main password when authenticating on the GitHub website. Everywhere else you should be using a PAT, preferably one that only has the permissions it needs.
+
+## Storing and loading your GitHub PAT
+
+Most R packages that interact with the GitHub API expect that your PAT is stored in the environment variable `GITHUB_PAT` in the R process. But how does it get there? The credentials package provides a function that will set this environment variable:
+
+```r
+# Tries to set the GITHUB_PAT environment variable
+credentials::set_github_pat()
+## TRUE
+```
+
+This function calls out to the [git credential store](https://git-scm.com/docs/git-credential-store) to get a suitable token for the `github.com` domain. If no token is available yet, the git credential manager will then prompt the user to enter one. What this looks like depends on your operating system and [credential helper](https://git-scm.com/docs/gitcredentials) configuration.
+
+![token](https://i.imgur.com/cnJcRmw.png)
+
```

Comment:
I'll try to improve that. But note that there is a bit of magic going on here because this prompt is invoked not by R, but by `git` when it calls back to the credential store. In some front-ends we don't have full control over the appearance of the password prompt (but in this example we do).